File: haskell/binary-search-tree/src/BST.hs
module BST
    ( BST
    , bstLeft
    , bstRight
    , bstValue
    , empty
    , fromList
    , insert
    , singleton
    , toList
    ) where

data BST a = Nil | Node (BST a) a (BST a) deriving (Eq, Show)

bstLeft :: BST a -> Maybe (BST a)
bstLeft Nil = Nothing
bstLeft (Node left _ _) = Just left 

bstRight :: BST a -> Maybe (BST a)
bstRight Nil = Nothing
bstRight (Node _ _ right) = Just right 

bstValue :: BST a -> Maybe a
bstValue Nil = Nothing
bstValue (Node _ value _) = Just value

empty :: BST a
empty = Nil

fromList :: Ord a => [a] -> BST a
fromList = foldl (flip insert) empty

insert :: Ord a => a -> BST a -> BST a
insert x Nil = singleton x
insert x (Node left value right)
  | x <= value = Node (insert x left) value right
  | otherwise = Node left value (insert x right)

singleton :: a -> BST a
singleton x = Node Nil x Nil

toList :: BST a -> [a]
toList Nil = []
toList (Node left value right) = (toList left) ++ [value] ++ (toList right)

File: scala/binary-search-tree/src/main/scala/Bst.scala
case class Bst[T](value: T, left: Option[Bst[T]] = None, right: Option[Bst[T]] = None)
(implicit evidence: T => Ordered[T]) {
  def insert(newValue: T): Bst[T] = {
    if (newValue > value)
      copy(right = right.map(_.insert(newValue)).orElse(Option(Bst(newValue))))
    else
      copy(left = left.map(_.insert(newValue)).orElse(Option(Bst(newValue))))
  }
}
object Bst {
  def fromList[T](list: List[T])(implicit evidence: T => Ordered[T]): Bst[T] = {
    list.tail.foldLeft(Bst(list.head)) { case (bst, value) => bst.insert(value) }
  }
  def toList[T](bst: Bst[T]): List[T] = {
    def loop(bstOption: Option[Bst[T]]): List[T] = bstOption match {
      case None => Nil
      case Some(bst) => loop(bst.left) ++ (bst.value :: loop(bst.right))
    }
    loop(Option(bst))
  }
}